src/array/cpu/id_hash_map.cc

```diff
@@ -0,0 +1,201 @@
+  // when the pointer is freed.
+  auto deleter = [](Mapping* mappings) {
+    if (mappings != nullptr) {
+      DGLContext ctx = DGLContext{kDGLCPU, 0};
+      auto device = DeviceAPI::Get(ctx);
+      device->FreeWorkspace(ctx, mappings);
+    }
+  };
+  hash_map_ = {nullptr, deleter};
+}
+
+template <typename IdType>
+IdArray IdHashMap<IdType>::Init(const IdArray ids) {
+  CHECK_EQ(ids.defined(), true);
+  const IdType* ids_data = ids.Ptr<IdType>();
+  const size_t num_ids = static_cast<size_t>(ids->shape[0]);
+  // Make sure `ids` is not 0 dim.
+  CHECK_GT(num_ids, 0);
+  size_t capcacity = GetMapSize(num_ids);
+  mask_ = static_cast<IdType>(capcacity - 1);
+
+  auto ctx = DGLContext{kDGLCPU, 0};
+  hash_map_.reset(static_cast<Mapping*>(
+      DeviceAPI::Get(ctx)->AllocWorkspace(ctx, sizeof(Mapping) * capcacity)));
+  memset(hash_map_.get(), -1, sizeof(Mapping) * capcacity);
+
+  // This code block is to fill the ids into hash_map_.
+  IdArray unique_ids = NewIdArray(num_ids, ctx, sizeof(IdType) * 8);
+  // Use `int16_t` instead of `bool`. As vector<bool> is an exception
+  // for whom updating different elements from different threads is unsafe.
+  // see https://en.cppreference.com/w/cpp/container#Thread_safety.
+  std::vector<int16_t> valid(num_ids);
+  auto thread_num = compute_num_threads(0, num_ids, kGrainSize);
+  std::vector<size_t> block_offset(thread_num + 1, 0);
+  IdType* unique_ids_data = unique_ids.Ptr<IdType>();
+
+  // Insert all elements in this loop.
+  parallel_for(0, num_ids, kGrainSize, [&](int64_t s, int64_t e) {
+    size_t count = 0;
+    for (int64_t i = s; i < e; i++) {
+      Insert(ids_data[i], &valid, i);
+      count += valid[i];
+    }
+    block_offset[omp_get_thread_num() + 1] = count;
+  });
+
+  // Get ExclusiveSum of each block.
+  std::partial_sum(
+      block_offset.begin() + 1, block_offset.end(), block_offset.begin() + 1);
+  unique_ids->shape[0] = block_offset.back();
+
+  // Get unique array from ids and set value for hash map.
+  parallel_for(0, num_ids, kGrainSize, [&](int64_t s, int64_t e) {
+    auto tid = omp_get_thread_num();
+    auto pos = block_offset[tid];
+    for (int64_t i = s; i < e; i++) {
+      if (valid[i]) {
+        unique_ids_data[pos] = ids_data[i];
+        Set(ids_data[i], pos);
+        pos = pos + 1;
+      }
+    }
+  });
+  return unique_ids;
+}
+
+template <typename IdType>
+IdArray IdHashMap<IdType>::MapIds(const IdArray ids) const {
+  CHECK_EQ(ids.defined(), true);
+  const IdType* ids_data = ids.Ptr<IdType>();
+  const size_t len = static_cast<size_t>(ids->shape[0]);
+  CHECK_GT(len, 0);
+
+  DGLContext ctx = DGLContext{kDGLCPU, 0};
+  IdArray new_ids = NewIdArray(len, ctx, sizeof(IdType) * 8);
+  IdType* values_data = new_ids.Ptr<IdType>();
+
+  parallel_for(0, len, kGrainSize, [&](int64_t s, int64_t e) {
+    for (int64_t i = s; i < e; i++) {
+      values_data[i] = MapId(ids_data[i]);
+    }
+  });
+  return new_ids;
+}
+
+template <typename IdType>
```

Comment:
not sure we need to explicitly code this, remove if it is unnecessary to be added.

Reply:
removed